Avoid double-free in PKCS8::load_key

hrantzsch opened the pull request:
This PR adds a test case to reproduce #2255 and a fix to avoid the issue.

 * [x] reproduce the issue on CI
 * [x] "fix" load_key to avoid the issue
 * [x] ~~run test only on clang 8 (or not?)~~

reneme commented on src/tests/test_clang_bug.cpp:
```diff
@@ -0,0 +1,119 @@
+         Botan::System_RNG rng;
+
+         auto pem = getPrivateKey();
+         std::string pw = "01234567890123456789"; // sufficiently long password
```
Sure that this is sufficiently long? On macOS I needed at least 23 bytes (including the `\0` byte, I believe).

hrantzsch replied:
It was enough in my local tests.
The strange thing is, testing the MVE you posted in the issue, very long passwords (~100 chars) didn't cause a crash but the issue was still visible in valgrind. 20 and 23 characters did cause a crash (double-free).

hrantzsch commented:
The bug was reproduced in the first commit, before adding the fix. [Here's the failed job](https://travis-ci.org/randombit/botan/jobs/651382999?utm_medium=notification&utm_source=github_status).

Regarding running the test only on the clang8 target, what's the best way to do this?
Also, I noticed that there already is a Valgrind target in the travis matrix. Should we also use `--with-valgrind` just to be sure, or is it enough that we've seen it crashing in the job linked above?

reneme commented on src/tests/test_clang_bug.cpp:
```diff
@@ -0,0 +1,125 @@
+            {
+            Botan::PKCS8::load_key(ds, rng, pw);
+            }
+         catch(const std::exception&) { }
```
Should we fail the test if `PKCS8::load_key` reports anything but "wrong password"?

hrantzsch replied:
Hm, initially I had it fail if there's no exception, but now I think that's not really what we want to test here. So I left it out to avoid confusion. I'm pretty sure there's another test that ensures this already.

reneme commented on src/tests/test_clang_bug.cpp:
```diff
@@ -0,0 +1,125 @@
+std::string getPrivateKey()
+   {
+   return
+      "-----BEGIN ENCRYPTED PRIVATE KEY-----"
```
For future reference: Do you remember the correct password? The test itself must use a wrong password, you mention that in the long description above. Future-me would find it helpful to see the "correct" password as another hint when quickly reading this test case

reneme commented:
> run test only on clang 8

Actually, I think it doesn't hurt to run this regression test on all compilers.

randombit commented:
Looks great thanks. I agree there is no reason to only run this under Clang 8 given the test should run very quickly. Always running it also ensures this never regresses under another compiler (at least not without us noticing..)

hrantzsch commented:
I changed the key for one where I still know the password and noted it, as @reneme suggested. Also, I changed my mind about checking whether `load_key` throws. The test now fails if it doesn't throw and succeeds only if it does.

All done from my side :)
